### PR TITLE
[GEOT-5392] Mosaicking refactor

### PR DESCRIPTION
**Pull request by dvntucker**
Lots of changes here. Refactored a lot of the Mosaicker into its own package. 

- GranuleCollector becomes SubmosaicProducer. SubmosaicProducer is now responsible for mosaicking its own contents.
- ReprojectingSubmosaicProducer implementation for OTF reprojection.
- All backwards compatible with imagemosaic
- Test cases for the new mosaic implementation

Mainly putting this PR together for ease of review/commenting

### CODE REVIEW

**Comment by aaime** [comment]
It seems there are conflicts

**Comment by dvntucker** [comment]
Yep, I'll take a look this morning. Hopefully nothing too crazy.

**Comment by aaime** [comment]
The first two commits create a lot of new public API (visible to anyone using Geotools, not just inside the mosaic) apparently only for the sake of moving 3 classes in their own package. I recommend moving those classes back and return the visibility to package private.

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/SubmosaicProducer.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+import org.geotools.gce.imagemosaic.GranuleDescriptor;
+import org.geotools.gce.imagemosaic.MosaicElement;
+import org.opengis.referencing.crs.CoordinateReferenceSystem;
+
+/**
+ * Responsible for creating subsets of the whole imagemosaic. Since there are frequently parts of
+ * a mosaic that need to be handled differently, a submosaic producer is responsible for handling
+ * those parts separately before they're added to the whole mosaic. For example a single dimension
+ * can be handled on its own before being
```
Comment (aaime) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/SubmosaicProducer.java]: "before being..." incomplete javadoc

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/SubmosaicProducer.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ */
+public interface SubmosaicProducer {
+    boolean accept(GranuleDescriptor granuleDescriptor);
+
+    List<MosaicElement> createMosaic() throws IOException;
+
+    boolean isDoInputTransparency();
+
+    boolean isHasAlpha();
```
Comment (aaime) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/SubmosaicProducer.java]: hasAlpha or isAlphaEnabled

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/SubmosaicProducer.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+import org.geotools.gce.imagemosaic.MosaicElement;
+import org.opengis.referencing.crs.CoordinateReferenceSystem;
+
+/**
+ * Responsible for creating subsets of the whole imagemosaic. Since there are frequently parts of
+ * a mosaic that need to be handled differently, a submosaic producer is responsible for handling
+ * those parts separately before they're added to the whole mosaic. For example a single dimension
+ * can be handled on its own before being
+ */
+public interface SubmosaicProducer {
```
Comment (aaime) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/SubmosaicProducer.java]: New public interface, no javadocs on its methods

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/SubmosaicProducer.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    boolean accept(GranuleDescriptor granuleDescriptor);
+
+    List<MosaicElement> createMosaic() throws IOException;
+
+    boolean isDoInputTransparency();
```
Comment (aaime) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/SubmosaicProducer.java]: isInputTransparencyEnabled maybe?

**Comment by dvntucker** [comment]
This is now rebased off of the formatting changes.

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/GranuleDescriptor.java [hunk 5]
```diff
@@ -137,14 +140,25 @@
 
     OverviewsController overviewsController;
 
+    private CoordinateReferenceSystem granuleCRS;
+
+    public GeneralEnvelope coverageEnvelope;
```
Comment (aaime) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/GranuleDescriptor.java]: The granule CRS is added and exposed via a getter, but this one is public?
The name is also confusing, what you really have is the granule/reader envelope, not the "coverage" one, indeed the descriptor does not deal with coverages at all.

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/GranuleDescriptor.java [hunk 6]
```diff
@@ -137,14 +140,25 @@
 
     OverviewsController overviewsController;
 
+    private CoordinateReferenceSystem granuleCRS;
```
Comment (aaime) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/GranuleDescriptor.java]: Looks like this one is a dead store, not initialized, nor used anywhere. The CRS information is also most likely available in the envelope (barring bugs, the envelope should always have a CRS, unless the reader does not have one at all)

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/GranuleDescriptor.java [hunk 7]
```diff
@@ -97,31 +98,33 @@
...
+    /**
+     * Logger.
+     */
     private final static Logger LOGGER = org.geotools.util.logging.Logging
             .getLogger(GranuleDescriptor.class);
 
     private static final String AUXFILE_EXT = ".aux.xml";
 
-    /** Hints to use for avoiding to search for the ImageMosaic format */
+    /**
```
Comment (aaime) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/GranuleDescriptor.java]: Looks like there still are lots of javadoc related pure reformats in this pull request. Please do not use automatic formatting, it makes bug fixes backports harder down the road :-)

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/GranuleDescriptor.java [hunk 8]
```diff
@@ -137,14 +140,25 @@
...
+    }
+
+    public void setGranuleCRS(CoordinateReferenceSystem granuleCRS) {
+        this.granuleCRS = granuleCRS;
+    }
+
     /**
      * This class represent an overview level in a single granuleDescriptor.
-     * 
+     * <p>
```
Comment (aaime) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/GranuleDescriptor.java]: Extra empty paragraph here

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/DefaultSubmosaicProducerFactory.java [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+import org.geotools.resources.coverage.FeatureUtilities;
+import org.opengis.filter.Filter;
+
+/**
+ * Create granule collectors based on whether the request was for a stacked mosaic with additional domains requested
+ */
+public class DefaultSubmosaicProducerFactory implements SubmosaicProducerFactory {
+
+    @Override
+    public List<SubmosaicProducer> createCollectors(RasterLayerRequest request,
```
Comment (aaime) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/DefaultSubmosaicProducerFactory.java]: Inconsistent naming, the javadoc and method name talks about collectors, but returns submosaic producers

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/ReprojectingSubmosaicProducer.java [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+        try {
+            granuleCRS = CRS.decode(granuleCRSCode);
+        } catch (FactoryException e) {
+            throw new IllegalStateException(e);
+        }
+
+        CRSBoundMosaicProducer submosaicProducer = this.perCrsMosaicProducers.get(granuleCRSCode);
+        if (submosaicProducer == null) {
+            submosaicProducer = new CRSBoundMosaicProducer(response, dryRun, granuleCRS);
+            perCrsMosaicProducers.put(granuleCRSCode, submosaicProducer);
```
Comment (aaime) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/ReprojectingSubmosaicProducer.java]: As discussed during the last voice synch up, this breaks sorting, the data structure should just be a list of crs bound producers, and every time a crs changes, a new bound producer should be added to the list

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/ReprojectingSubmosaicProducerFactory.java [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+import org.geotools.gce.imagemosaic.RasterLayerRequest;
+import org.geotools.gce.imagemosaic.RasterLayerResponse;
+import org.geotools.gce.imagemosaic.RasterManager;
+
+/**
+ * Creates a SubmosaicProducer that can handle reprojecting granules into the mosaic's target CRS
+ */
+public class ReprojectingSubmosaicProducerFactory implements SubmosaicProducerFactory {
+    @Override
+    public List<SubmosaicProducer> createCollectors(RasterLayerRequest request,
```
Comment (aaime) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/ReprojectingSubmosaicProducerFactory.java]: Makes it look like heterogeneous crs mosaic are not going to support stacking merge behavior on ND mosaics correctly (since the default  submosaic producer has special handling for it and will return multiple submosaic producers for it, but not this one)

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/properties/CRSExtractor.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    public static final String DEFAULT_ATTRIBUTE_NAME = "crs";
+
+    private String crsCode;
+
+    public CRSExtractor() {
+        this(new CRSExtractorSPI(), Collections.singletonList(DEFAULT_ATTRIBUTE_NAME));
+    }
+
+    public CRSExtractor(PropertiesCollectorSPI spi, List<String> propertyNames) {
+        super(spi, propertyNames);
```
Comment (aaime) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/properties/CRSExtractor.java]: The list of property names is provided because other extractors are setup with regexes and can extract multiple values (groups) in one shot, but for this one, I'd actually throw an exception if there are multiple properties, what purpose does it serve to collect the same crs value multiple times in different attributes? The same goes for DateExtractor, FSDateExtractor, they could use a "SingleValuePropertiesCollector" base class, or something like that.

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/catalog/GranuleCatalog.java [hunk 13]
```diff
@@ -147,4 +151,12 @@ protected MultiLevelROI getGranuleFootprint(SimpleFeature sf) {
      * @throws IOException in case something bad happens
      */
     public abstract void drop() throws IOException;
+
+    public SubmosaicProducerFactory getCollectorsFactory() {
```
Comment (aaime) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/catalog/GranuleCatalog.java]: It is not the job of the granule catalog to know about the collectors/mosaic producers, it is just to know about the granule originators and descriptors, but not about how to build the results. The raster manager should pick it from the MosaicConfigurationBean on initialization instead, and probably make it a field to avoid re-instantiating it over and over.

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/SubmosaicProducer.java [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+import org.geotools.gce.imagemosaic.GranuleDescriptor;
+import org.geotools.gce.imagemosaic.MosaicElement;
+import org.opengis.referencing.crs.CoordinateReferenceSystem;
+
+/**
+ * Responsible for creating subsets of the whole imagemosaic. Since there are frequently parts of a mosaic that need to be handled differently, a
+ * submosaic producer is responsible for handling those parts separately before they're added to the whole mosaic. For example a single dimension can
+ * be handled on its own before being
+ */
+public interface SubmosaicProducer {
```
Comment (aaime) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/SubmosaicProducer.java]: So this new interface takes over the responsibiliites of both GranuleCollector and MosaicProducer, mixing them in a single class?

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/RasterLayerResponse.java [hunk 15]
```diff
@@ -259,410 +252,11 @@ public ColorInterpretation getColorInterpretation() {
...
-
-    /**
      * A class doing the mosaic operation on top of a List of {@link MosaicElement}s.
-     * 
-     * @author Simone Giannecchini, GeoSolutions SAS
      *
+     * @author Simone Giannecchini, GeoSolutions SAS
      */
-    private class Mosaicker {
+    public static class Mosaicker {
```
Comment (aaime) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/RasterLayerResponse.java]: If it's public/static it could well go in its own file now

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/RasterLayerResponse.java [hunk 16]
```diff
@@ -875,150 +511,74 @@ private MosaicElement createMosaic() throws IOException {
 
     /**
      * This class is responsible for putting together the granules for the final mosaic.
-     * 
+     *
      * @author Simone Giannecchini, GeoSolutions SAS
-     * 
      */
     private class MosaicProducer implements GranuleCatalogVisitor {
```
Comment (aaime) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/RasterLayerResponse.java]: Looks like this class has no use anymore? I don't see it being used/instantiated anywhere.

Comment (aaime) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/RasterLayerResponse.java]: Never mind, found a usage in RasterLayerResponse.prepareResponse, the search output in Eclipse looked it preparedResponse was a method of MosaicProdurcer instead.

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/ReprojectingSubmosaicProducerFactory.java (hunk 11 above)
Comment (aaime) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/ReprojectingSubmosaicProducerFactory.java]: Btw, when this is done, make sure to also add a last sortby on the crs field to avoid jumps across different CRSs in granules having the same values for the main sort fields (e.g., if sorting on time to have most recent on top, make sure features with the same date are not in mixed crs order, but packed instead)

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/DefaultSubmosaicProducerFactory.java [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+            RasterManager rasterManager, RasterLayerResponse response, boolean dryRun) {
+        // get merge behavior as per request
+        List<SubmosaicProducer> defaultSubmosaicProducers = new ArrayList<>();
+        MergeBehavior mergeBehavior = request.getMergeBehavior();
+
+        // prepare dimensions management if needed, that is in case we use stacking
+        if (mergeBehavior.equals(MergeBehavior.STACK)) {
```
Comment (aaime) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/DefaultSubmosaicProducerFactory.java]: This code is orthogonal to the normal collection strategies and reprojection itself, it's designed to allow supporting WMS-EO, in which a mosaic has N bands split as separate granules, that need to be selected, collected and mosaicked separately, and then band-merged to return the custom band composition requested in the WMS-EO extensions.

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/ReprojectingSubmosaicProducer.java (hunk 10 above)
Comment (dvntucker) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/ReprojectingSubmosaicProducer.java]: I have a fix for the SortBy stuff mostly ready, I just need to do some more testing and create a test case.

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/GranuleDescriptor.java (hunk 5 above)
Comment (dvntucker) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/GranuleDescriptor.java]: I'll change the name to be more accurate and take a look at the visibility to make sure it's correct.

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/GranuleDescriptor.java (hunk 6 above)
Comment (dvntucker) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/GranuleDescriptor.java]: Ok, I'll take a look at getting the CRS from the Envelope instead. Probably a better place to get it.

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/GranuleDescriptor.java (hunk 7 above)
Comment (dvntucker) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/GranuleDescriptor.java]: I'll take a look at fixing some of these. Just an unfortunate side effect of the differences between IntelliJ/Eclipse formatting. I reformatted my entire branch with Eclipse; however some of the multiline comments had been expanded by IntelliJ already and for some reason the Eclipse formatter won't change the format of existing comments.

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/DefaultSubmosaicProducerFactory.java (hunk 9 above)
Comment (dvntucker) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/DefaultSubmosaicProducerFactory.java]: I'll fix up the JavaDoc. Some of these were just automatically refactored from existing classes.

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/ReprojectingSubmosaicProducerFactory.java (hunk 11 above)
Comment (dvntucker) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/ReprojectingSubmosaicProducerFactory.java]: I'm not sure we're going the resources to support a stacked mosaic in this. I'm going to take a look at what it would take, maybe I can do it fairly simply. We may have additional time/money to tackle this in September though. Hopefully it's not a deal-breaker, since stacking merge is still available in the default mosaic config.

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/properties/CRSExtractor.java (hunk 12 above)
Comment (dvntucker) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/properties/CRSExtractor.java]: I can make a change for that.

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/catalog/GranuleCatalog.java (hunk 13 above)
Comment (dvntucker) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/catalog/GranuleCatalog.java]: I'll see if I can move this somewhere where it makes sense.

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/SubmosaicProducer.java (hunk 14 above)
Comment (dvntucker) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/SubmosaicProducer.java]: That's basically right yes. So that the new class can collect information needed to handle granule groups that may need special handling (such as reprojecting).

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/RasterLayerResponse.java (hunk 15 above)
Comment (dvntucker) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/RasterLayerResponse.java]: I can refactor out. I think there are some other inner classes that can be moved to the outer level now.

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/DefaultSubmosaicProducerFactory.java (hunk 17 above)
Comment (dvntucker) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/DefaultSubmosaicProducerFactory.java]: Maybe I should move that code into another specific subclass that's only used if the STACK behavior is used?

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/ReprojectingSubmosaicProducer.java (hunk 10 above)
Comment (dvntucker) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/ReprojectingSubmosaicProducer.java]: SortBy is respected now. I do this by grouping by every CRS + unique sort value. This could be fairly inefficient if the sort is by a large continuous range of values, but I think in many cases people will be sorting by a smaller number of discrete values (say, resolution or dates)

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/properties/CRSExtractor.java (hunk 12 above)
Comment (dvntucker) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/properties/CRSExtractor.java]: I'm not sure if we're going to have time to make this change in this PR. It's slightly more involved because PropertyCollector itself makes assumptions about having a list of properties. I think ideally PropertyCollector would be an interface, or PropertyCollector itself would be even more generic and have different Multi/Single value implementations.

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/GranuleDescriptor.java (hunk 5 above)
Comment (dvntucker) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/GranuleDescriptor.java]: This has been renamed and been made private.

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/catalog/GranuleCatalog.java (hunk 13 above)
Comment (dvntucker) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/catalog/GranuleCatalog.java]: Moved all of this to RasterManager, removed all references in the Catalog code.

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/properties/CRSExtractor.java (hunk 12 above)
Comment (aaime) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/properties/CRSExtractor.java]: No problem, was not a big deal

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/ReprojectingSubmosaicProducer.java [hunk 18]
```diff
@@ -158,13 +206,52 @@ public CRSBoundMosaicProducer(RasterLayerResponse rasterLayerResponse, boolean d
...
+                    + "Cannot proceed with mosaicking.", e);
+            }
+
+            //we need to see if the incoming granule matches the "template" feature this producer
+            //was initialized with.
+            if (sortBy != null) {
+                SimpleFeature granuleFeature = granuleDescriptor.getOriginator();
+                SimpleFeature templateFeature = templateDescriptor.getOriginator();
+                for (SortBy sortKey : sortBy) {
+                    Object incomingAttribute = granuleFeature
```
Comment (aaime) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/ReprojectingSubmosaicProducer.java]: I'm not sure I understand why this is necessary. I though ensuring uniform CRS in the producer was enough. E.g., if I have all features in the same CRS but I'm sorting on date, do I really need to roll each feature in its own sub-mosaicker?

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/RasterLayerResponse.java [hunk 19]
```diff
@@ -1174,10 +1176,11 @@ private void handleSortByClause(final Query query) {
...
 
                 // assign to query if sorting is supported!
-                final SortBy[] sb = clauses.toArray(new SortBy[] {});
+
+                this.sortBy = clauses.toArray(new SortBy[] {});
                 if (rasterManager.getGranuleCatalog()
-                        .getQueryCapabilities(rasterManager.getTypeName()).supportsSorting(sb)) {
-                    query.setSortBy(sb);
+                        .getQueryCapabilities(rasterManager.getTypeName()).supportsSorting(sortBy)) {
+                    query.setSortBy(sortBy);
```
Comment (aaime) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/RasterLayerResponse.java]: In case of heterogeneous CRS I'd add the crs property as a last sort-by property, so that we don't end up creating sub-mosaics for nothing (mosaicking is an expensive operation, we should try to do as few as possible within the respect of the user request)

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/DefaultSubmosaicProducerFactory.java (hunk 17 above)
Comment (aaime) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/granulecollector/DefaultSubmosaicProducerFactory.java]: Given that we plan to refactor the mosaic in the future to simplify its structure further, I'd say to let this one go for the moment. We are also considering moving the stacking approach up in the food chain and have it implemented purely in GeoServer later (as it's only used by WMS-EO at the moment).

**Comment by aaime** [comment]
Just added a few more minor comments, everything looks otherwise pretty good. Looks like there are merge conflicts though.

**Comment by dvntucker** [comment]
I'll squash and rebase. I also want to rename one of the packages (granule collector) before merging.

File: modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/RasterLayerResponse.java (hunk 19 above)
Comment (dvntucker) [inline modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/RasterLayerResponse.java]: I think we're going to leave this for the user to control with the SortBy parameter for now. In the future I'd like to update maybe SubmosaicProducerFactory to have an optional SortBy parameter so that the RasterLayerRequest can query.

**Comment by dvntucker** [comment]
Simplified the Sorting along the lines we discussed last week.

**Comment by bmmpxf** [comment]
@aaime Is this ready to be merged now?
